### PR TITLE
TCP comms. with unittesting

### PR DESCRIPTION
**Pull request by caryan**
Provides TCP comms with fallback to UDP on older firmware along with fairly complete communications unittests.

### CODE REVIEW

File: src/C++/enumerate.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    cout << concol::CYAN << "Device " << ct << " at IPv4 address " << serialBuffer[ct] <<
+		" running firmware version " << print_firmware_version(firmware_version) <<
+		" has been up " << uptime_pretty.str() << concol::RESET << endl;
+		disconnect_APS(serialBuffer[ct]);
+  }
+	cout << endl;
+
+  return 0;
+}
```
Comment (blakejohnson) [inline src/C++/enumerate.cpp]: This file has inconsistent indentation. Please go all tabs or all spaces.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -80,10 +90,14 @@ ADD_EXECUTABLE(play_sequence
...
-	./util/test_comms.cpp
+ADD_EXECUTABLE(run_tests
+	../test/run_tests.cpp
+	../test/test_Sanity.cpp
+	../test/test_datagram.cpp
+	./lib/APS2Datagram.cpp
+	../test/test_comms.cpp
 )
-TARGET_LINK_LIBRARIES(test_comms aps2)
+TARGET_LINK_LIBRARIES(run_tests aps2 pthread)
```
Comment (blakejohnson) [inline src/CMakeLists.txt]: `pthread` is a linux-only dependency.

File: src/lib/APS2.cpp [hunk 3]
```diff
@@ -1,20 +1,27 @@
...
+void APS2::connect(shared_ptr<APS2Ethernet> && ethernetRM) {
+	FILE_LOG(logDEBUG) << "APS2::connect";
+	//Hold on to APS2Ethernet class to keep socket alive
 	ethernetRM_ = ethernetRM;
 	if (!isOpen) {
 		ethernetRM_->connect(deviceSerial_);
+		//Figure out whether this is an APS2 or TDM and what register map to use
 		try {
 			APSStatusBank_t statusRegs = read_status_registers();
+			if ( statusRegs.userFirmwareVersion != 0xbadda555 ) {
```
Comment (blakejohnson) [inline src/lib/APS2.cpp]: cute

Comment (blakejohnson) [inline src/lib/APS2.cpp]: But I have sometimes shoved random stuff into that register for debugging purposes. Can we just check the major firmware version number (assuming you rolled to 3.x...)?

File: src/lib/APS2.cpp [hunk 4]
```diff
@@ -44,41 +51,31 @@ void APS2::disconnect() {
...
-		} catch (std::exception &e) {
-			FILE_LOG(logDEBUG) << "Status timeout; retrying...";
-		}
-		retrycnt++;
-	}
+	//Send the command
+	ethernetRM_->send(deviceSerial_, {{cmd, addr, {}}});
 
-	throw APS2_RESET_TIMEOUT;
+	//we expect to loose the connection at this point...
```
Comment (blakejohnson) [inline src/lib/APS2.cpp]: A note to myself: check how Colm looks for a successful reset.

File: src/lib/APS2.cpp [hunk 5]
```diff
@@ -178,82 +211,90 @@ double APS2::get_fpga_temperature(){
...
-	vector<uint32_t> packedData;
-	packedData.resize(fileData.size()/4);
-	memcpy(packedData.data(), fileData.data(), fileData.size());
-
-	//Convert to big endian byte order - basically because it will be byte-swapped again when the packet is serialized
-	for (auto & packet : packedData) {
-		packet = htonl(packet);
+	//Get the file size in bytes
+	std::ifstream FID (bitFile, std::ios::in|std::ios::binary);
+  if (!FID.is_open()){
```
Comment (blakejohnson) [inline src/lib/APS2.cpp]: oops, indentation switched to spaces.

File: src/lib/APS2.cpp [hunk 6]
```diff
@@ -178,82 +211,90 @@ double APS2::get_fpga_temperature(){
...
+			break;
+	}
 
-	return ethernetRM_->send(deviceSerial_, packet, false);
+	//Now validate bitfile data (EPROM only for now)
+	//TODO: DRAM as well?
+	flash_validate_done = 0;
+	if (media == BITFILE_MEDIA_EPROM) {
+		uint32_t end_addr = start_addr + 4*bitfile_words.size();
+		uint32_t addr{start_addr};
```
Comment (blakejohnson) [inline src/lib/APS2.cpp]: I find this style of variable initialization to be less clear than `uint32_t addr = start_addr`.

File: src/C++/helpers.h [hunk 7]
```diff
@@ -62,3 +62,23 @@ void print_title(const string & title){
...
   cout << endl;
 }
+
+//Copy of APS2::print_firmware_version
+string print_firmware_version(uint32_t version_reg) {
+	std::ostringstream ret;
+
+	uint32_t minor{version_reg & 0xff};
+	uint32_t major{(version_reg >> 8) & 0xf};
+	uint32_t note_nibble{(version_reg >> 12) & 0xf};
```
Comment (blakejohnson) [inline src/C++/helpers.h]: Can we use simple assignment initialization for these variables? For bitstypes, bracket initialization seems less readable to me.

File: src/lib/APS2.cpp [hunk 8]
```diff
@@ -487,58 +528,98 @@ void APS2::write_memory(const uint32_t & addr, const vector<uint32_t> & data){
 	 * data = vector<uint32_t> data
 	 */
 
-	if (data.size() == 0) {
-		return;
-	}
```
Comment (blakejohnson) [inline src/lib/APS2.cpp]: This empty packet check was important for a certain edge case on the old firmware. Does the unit test suite contain an empty message check?

File: src/lib/APS2.cpp [hunk 9]
```diff
@@ -1041,28 +1083,31 @@ void APS2::check_clocks_status() {
...
-	 	for (auto bit : {MMCM_SYS_LOCK_BIT, MMCM_CFG_LOCK_BIT, MIG_C0_LOCK_BIT, MIG_C0_CAL_BIT, MIG_C1_LOCK_BIT, MIG_C1_CAL_BIT}) {
-	 		clocksGood &= (statusRegs.userStatus >> bit) & 0x1;
-	 	}
-	 	if (!clocksGood) {
-	 		throw APS2_MMCM_LOST_LOCK;
-	 	}
+	//On board MMCM PLL's are reported either through user status or the PLL_STATUS_ADDR CSR register
+	uint32_t pll_status;
+	if (legacy_firmware) {
+		if (statusRegs.userFirmwareVersion < 0x212) return; //nothing to read
```
Comment (blakejohnson) [inline src/lib/APS2.cpp]: I suppose we should put this special firmware version in a `#define` for legibility.

Comment (blakejohnson) [inline src/lib/APS2.cpp]: I am also tempted to just drop this check since I don't think any firmware versions that old exist in the wild any longer.

File: src/lib/APS2.cpp [hunk 10]
```diff
@@ -1083,11 +1128,11 @@ int APS2::get_PLL_freq() {
 			case 0x11: freq = 300; break;
 			case 0x00: freq = 600; break;
 			default:
-				return -2;
+				return -1;
```
Comment (blakejohnson) [inline src/lib/APS2.cpp]: should this throw an error rather than returning a magic number?

File: src/lib/APS2.cpp [hunk 11]
```diff
@@ -1541,7 +1577,7 @@ int APS2::set_offset_register(const int & dac, const float & offset) {
 }
 
 void APS2::set_bit(const uint32_t & addr, std::initializer_list<int> bits) {
-	uint32_t curReg = read_memory(addr, 1)[0];
+	uint32_t curReg = read_memory(addr, 1).front();
```
Comment (blakejohnson) [inline src/lib/APS2.cpp]: gratuitous change :)

File: src/lib/APS2Datagram.cpp [hunk 12]
```diff
@@ -0,0 +1,113 @@
+// Datagrams that are sent/received from the APS2
+// Consists of a vector of 32 bit words
+// 1. command word - see APS2Command
+// 2. address
+// 3. payload
+//
+// Original author: Colm Ryan
+// Copyright 2016 Raytheon BBN Technologi
```
Comment (blakejohnson) [inline src/lib/APS2Datagram.cpp]: Technologies

File: src/lib/APS2Datagram.cpp [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+	//Create multiple datagrams of a maximum payload size
+	vector<APS2Datagram> chunks;
+	auto start_it = data.begin();
+	while (start_it != data.end()) {
+		auto end_it = (std::distance(start_it, data.end()) > chunk_size) ? start_it + chunk_size : data.end();
+		auto cur_chunk_size = std::distance(start_it, end_it);
+		cmd.cnt = cur_chunk_size;
+		chunks.push_back(APS2Datagram{cmd, addr, vector<uint32_t>(start_it, end_it)});
+		std::advance(start_it, cur_chunk_size);
+		addr += 4*cur_chunk_size;
```
Comment (blakejohnson) [inline src/lib/APS2Datagram.cpp]: instead of `4` maybe `sizeof(uint32_t)`?

File: src/lib/APS2Datagram.cpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+	//Construct the datagram
+	vector<uint32_t> data;
+	data.reserve(2+payload.size());
+	data.push_back(cmd.packed);
+	data.push_back(addr);
+	std::copy(payload.begin(), payload.end(), std::back_inserter(data));
+	return data;
+}
+
+void APS2Datagram::check_ack(const APS2Datagram & ack, bool legacy_firmware) const {
```
Comment (blakejohnson) [inline src/lib/APS2Datagram.cpp]: this is a nice improvement... do all the ACK checking in one place so it moves out of "user" code.

File: src/lib/APS2Datagram.h [hunk 15]
```diff
@@ -38,37 +49,24 @@ typedef union {
...
-class APSEthernetPacket{
+class APS2Datagram {
+private:
+  /* data */
 public:
-	APSEthernetHeader header;
+  APS2Command cmd;
+  uint32_t addr;
 	vector<uint32_t> payload;
+	vector<uint32_t> data() const;
```
Comment (blakejohnson) [inline src/lib/APS2Datagram.h]: indentation switch

File: src/lib/APS2Ethernet.cpp [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+        udp_socket_old_.open(udp::v4());
+        udp_socket_old_.bind(udp::endpoint(udp::v4(), APS_PROTO));
+    } catch (...) {
+        throw APS2_SOCKET_FAILURE;
+    }
+
+    //Bind UDP socket at local port bb4f
+    try {
+        udp_socket_.open(udp::v4());
+        udp_socket_.bind(udp::endpoint(udp::v4(), APS_PROTO+1));
```
Comment (blakejohnson) [inline src/lib/APS2Ethernet.cpp]: Just to confirm: the future plan here is to not bind to any particular local socket and have the APS2 respond back to whichever socket the message came from, right?

(I understand that the firmware does not do this, yet).

File: src/lib/APS2Ethernet.cpp [hunk 17]
```diff
@@ -0,0 +1,560 @@
...
+                vector<uint8_t> packetData(received_udp_data_old_, received_udp_data_old_ + bytesReceived);
+                sorter_lock_.lock();
+                sort_packet(packetData, remote_udp_endpoint_old_);
+                sorter_lock_.unlock();
+            }
+
+            //Start the receiver again
+            setup_udp_receive_old();
+    });
+}
```
Comment (blakejohnson) [inline src/lib/APS2Ethernet.cpp]: These two methods are the same except of the values `received_udp_data_old_` and `remote_udp_endpoint_old_`, no? So, can we make those arguments to the function, instead?

File: src/lib/APS2Ethernet.cpp [hunk 18]
```diff
@@ -0,0 +1,560 @@
...
+        }
+    }
+    freeifaddrs(ifap);
+    #endif //_WIN32
+
+    return IPs;
+}
+
+set<string> APS2Ethernet::enumerate() {
+	/*
```
Comment (blakejohnson) [inline src/lib/APS2Ethernet.cpp]: inconsistent indentation in this method.

File: src/lib/APS2Ethernet.cpp [hunk 19]
```diff
@@ -0,0 +1,560 @@
...
+    udp_socket_old_.send_to(asio::buffer(broadcastPacket.serialize()), endpoint);
+    //And the new
+    endpoint.port(UDP_PORT);
+    uint8_t enumerate_request = 0x01;
+    udp_socket_.send_to(asio::buffer(&enumerate_request, 1), endpoint);
+
+    //Wait 100ms for response
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+    //Check again
```
Comment (blakejohnson) [inline src/lib/APS2Ethernet.cpp]: comment seems out of date.

File: src/lib/APS2Ethernet.cpp [hunk 20]
```diff
@@ -0,0 +1,560 @@
...
+    std::unique_ptr<tcp::socket> sock(new tcp::socket(ios_));
+    asio::error_code ec;
+    FILE_LOG(logDEBUG1) << "Trying to connect to TCP port";
+    sock->connect(tcp::endpoint(asio::ip::address_v4::from_string(serial), TCP_PORT), ec);
+    if (ec) {
+      FILE_LOG(logERROR) << "Failed to connect to " << serial << " with error: " << ec.message();
+      throw APS2_SOCKET_FAILURE;
+    }
+
+    tcp_sockets_.insert(std::make_pair(serial, std::move(sock)));
```
Comment (blakejohnson) [inline src/lib/APS2Ethernet.cpp]: I bet it took a while to find that incantation...

File: src/lib/APS2Ethernet.cpp [hunk 21]
```diff
@@ -0,0 +1,560 @@
...
+        //Apply acknowledge flag to last element of chunk
+        if (!noACK) {
+            buffer.back().header.command.cmd &= ~(1 << 3);
+        }
+
+        send_chunk(serial, buffer, noACK);
+        std::advance(iter, chunkSize);
+
+        if (verbose && (std::distance(msg.begin(), iter) % 1000 == 0)) {
+            FILE_LOG(logDEBUG) << "Write " << 100*std::distance(msg.begin(), iter)/msg.size() << "% complete";
```
Comment (blakejohnson) [inline src/lib/APS2Ethernet.cpp]: These messages don't interfere with your status messages in the caller?

File: src/lib/APS2Ethernet.cpp [hunk 22]
```diff
@@ -0,0 +1,560 @@
...
+    // If we have TCP just send it out
+    for (const auto & dg : datagrams){
+      auto data = dg.data();
+      for (auto & val : data) {
+        val = htonl(val);
+      }
+      FILE_LOG(logDEBUG3) << "Sending datagram with command word " << hexn<8> << dg.cmd.packed <<
+        " to address " << hexn<8> << dg.addr << " with payload size " << std::dec << dg.payload.size() << " for total size " << data.size();
+      std::error_code ec;
+      asio::write(*tcp_sockets_[ipAddr], asio::buffer(data), ec); //TODO should this be async?
```
Comment (blakejohnson) [inline src/lib/APS2Ethernet.cpp]: I think you get "async-ness" for free, here.

File: src/lib/APS2Ethernet.cpp [hunk 23]
```diff
@@ -0,0 +1,560 @@
...
+
+    auto read_with_timeout = [&]() {
+      std::future<size_t> read_result = tcp_sockets_[ipAddr]->async_receive(asio::buffer(buf), asio::use_future);
+      if (read_result.wait_for(timeout) == std::future_status::timeout) {
+        FILE_LOG(logERROR) << "TCP receive timed out!";
+        throw APS2_RECEIVE_TIMEOUT;
+      }
+    };
+
+	  //First read header (cmd and addr)
```
Comment (blakejohnson) [inline src/lib/APS2Ethernet.cpp]: comment indentation

File: test/test_comms.cpp [hunk 24]
```diff
@@ -0,0 +1,312 @@
...
+		REQUIRE ( temp < 70 );
+	}
+
+	SECTION("variable write sizes") {
+		//Variable write sizes from 4 to 256 words
+		vector<uint32_t> outData, inData;
+		size_t idx=0;
+		std::default_random_engine generator;
+		std::uniform_int_distribution<uint32_t> wordDistribution;
+		for (unsigned writeSize=4; writeSize<=256; writeSize+=4){
```
Comment (blakejohnson) [inline test/test_comms.cpp]: Should start with `writeSize = 0`.

**Comment by blakejohnson** [comment]
LOVE the new unit testing.

**Comment by blakejohnson** [comment]
:100: Looks *really* good. I made a lot of comments, but they are all fairly minor.

File: src/lib/APS2.cpp (hunk 8 above)
Comment (blakejohnson) [inline src/lib/APS2.cpp]: I just checked and the unit tests do not currently include the empty message case. We should fix that and then revisit this.

File: src/lib/APS2.cpp (hunk 3 above)
Comment (caryan) [inline src/lib/APS2.cpp]: The firmware version has moved to it's own register in the CSR.   There's still the top 16 bits available but I was actually tempted to use them for the start of the git SHA-1 for unstable releases (kind of like Julia).   Then we could have less tagged releases but still mark each bitfile and know what it was able to do.

File: src/lib/APS2.cpp (hunk 4 above)
Comment (caryan) [inline src/lib/APS2.cpp]: Uh, he doesn't....

As you know the issue is what happens if the IP address changes. I was thinking of enumerating just before resetting and then renumerating and jumping if out IP has disappeared and a new one has appeared.

File: src/C++/helpers.h (hunk 7 above)
Comment (caryan) [inline src/C++/helpers.h]: Yep, I can't really disagree with you.  It allows the compiler to catch some unintended narrowing but that's usually more annoying than helpful.

File: src/lib/APS2.cpp (hunk 8 above)
Comment (caryan) [inline src/lib/APS2.cpp]: Hmmm, we should add coverage because I think it will break the new firmware too.   0 bytes to transfer DataMover commands are illegal.

File: src/lib/APS2Datagram.cpp (hunk 13 above)
Comment (caryan) [inline src/lib/APS2Datagram.cpp]: Yeah, you're right.  I though the fixed width integers guaranteed their width but it's only  the binary representation that's guaranteed not the storage.

File: src/lib/APS2Ethernet.cpp (hunk 16 above)
Comment (caryan) [inline src/lib/APS2Ethernet.cpp]: Yeah, I think that's the right way to do it and should allow multiple libaps2 instances.

File: src/lib/APS2Ethernet.cpp (hunk 17 above)
Comment (caryan) [inline src/lib/APS2Ethernet.cpp]: Yes, that should clean up nicely.

File: src/lib/APS2Ethernet.cpp (hunk 20 above)
Comment (caryan) [inline src/lib/APS2Ethernet.cpp]: Ha! It did.   That's why I put off making APS2 non-copyable.

File: src/lib/APS2Ethernet.cpp (hunk 22 above)
Comment (caryan) [inline src/lib/APS2Ethernet.cpp]: `asio::write` is explicitly blocking and blocking till all the data is sent too.  But it actually would be a hassle to async send because you  have to keep the send buffer valid until the callback.

File: src/lib/APS2Ethernet.cpp (hunk 21 above)
Comment (caryan) [inline src/lib/APS2Ethernet.cpp]: I have nice progress bars now so `program.cpp` no longer logs to `stdout`.

File: src/lib/APS2.cpp (hunk 3 above)
Comment (blakejohnson) [inline src/lib/APS2.cpp]: Ah, I really like the idea of putting the git SHA in the register. We'll have to figure out a way to make that automatic.

File: src/lib/APS2Datagram.cpp (hunk 13 above)
Comment (blakejohnson) [inline src/lib/APS2Datagram.cpp]: I prefer `sizeof(uint32_t)` just for readability sake, since it makes it obvious where the `4` is coming from.

File: src/lib/APS2.cpp (hunk 8 above)
Comment (caryan) [inline src/lib/APS2.cpp]: We're actually okay.  We're saved by the datagram chunking returning an empty vector and then there is nothing to do. Added test coverage for both in 0375f64c44d9babb980f6ad55f7462e46bcbab9c

File: src/lib/APS2Datagram.cpp (hunk 13 above)
Comment (caryan) [inline src/lib/APS2Datagram.cpp]: On closer inspection we do want the raw `4`.  This is incrementing the AXI byte address and irrespective of the host storage it has to be 4.

File: src/lib/APS2.cpp (hunk 10 above)
Comment (caryan) [inline src/lib/APS2.cpp]: Done in eda6c7aa

File: src/lib/APS2.cpp (hunk 9 above)
Comment (caryan) [inline src/lib/APS2.cpp]: Dropped in 44369a14

File: src/CMakeLists.txt (hunk 2 above)
Comment (blakejohnson) [inline src/CMakeLists.txt]: Sorry, I take that back. `pthread` should work on both `gcc` and `clang`, so we can keep it here.

File: src/lib/APS2Datagram.cpp (hunk 13 above)
Comment (blakejohnson) [inline src/lib/APS2Datagram.cpp]: Ok.

File: src/lib/APS2Ethernet.cpp (hunk 17 above)
Comment (caryan) [inline src/lib/APS2Ethernet.cpp]: Also have to pass and capture the socket and buffer but done in a211c71
